### PR TITLE
show gas estimation on buy

### LINKED ISSUE #1066: Show gas cost estimation for buy
We can display how much the tx will cost to order ( or buy dt + order for pool).  So somewhere in the use tab we can display this.

### PR DESCRIPTION
**Pull request by EnzoVezzaro**
- Fixes #1066 

Changes proposed in this PR:

- added gas estimation to buy

### CODE REVIEW

File: src/@utils/order.ts [hunk 1]
```diff
@@ -48,57 +47,153 @@ export async function order(
...
 
-  // TODO: we need to approve provider fee
+  return {
+    initializeData,
+    orderParams
+  }
+}
+
+/**
+ * Before stating the transactions, we need to initialized our provider's instance
```
Comment (claudiaHash) [inline src/@utils/order.ts:57]: Here are some typos which we should correct before merging

File: src/@utils/ocean.ts [hunk 2]
```diff
@@ -72,3 +74,23 @@ export async function getOceanBalance(
     LoggerInstance.error(`ERROR: Failed to get the balance: ${e.message}`)
   }
 }
+
+export async function convertGasFeesToOcean(
```
Comment (kremalicious) [inline src/@utils/ocean.ts:78]: not sure we have to do all this conversion as making our `usePrices()` hook also deliver ETH prices is super trivial as the Coingecko API allows passing multiple token IDs, so would stay just one request. But guess we would have to refactor some places then

File: src/@utils/order.ts [hunk 3]
```diff
@@ -48,57 +47,153 @@ export async function order(
...
+/**
+ * Get and execute transaction based on asset type
+ * @param web3
+ * @param asset
+ * @param accountId
+ * @param action
+ * @param orderPriceAndFees
+ * @returns {TransactionReceipt} receipt of the order / gas estimate
+ */
+async function getTransaction(
```
Comment (kremalicious) [inline src/@utils/order.ts:65]: good direction centralizing this, just a note that we need to test all possible download options with this change as PR touches all of them.

Just logically, the estimations are no transactions so not sure this is the best place to hold them too

File: src/@utils/ocean.ts (hunk 2 above)
Comment (EnzoVezzaro) [reply src/@utils/ocean.ts:78]: Ok, I will look into it. At this level also, we'll need to make some adjustments based on changes to [`Prices.tsx`](https://github.com/oceanprotocol/market/pull/1132/files). We'll need to add the fix [here](https://github.com/oceanprotocol/market/blob/feature/issue-1066-gas-estimation-buy/src/components/Asset/AssetActions/Download.tsx#L72:L72) to make sure these calculations are done for matic too.

File: src/components/Asset/AssetActions/Download.tsx [hunk 4]
```diff
@@ -53,16 +63,42 @@ export default function Download({
     setValidOrderTx(asset?.accessDetails?.validOrderTx)
     // get full price and fees
     async function init() {
+      // we only show the gas estimation if the user is connected
```
Comment (kremalicious) [inline src/components/Asset/AssetActions/Download.tsx:66]: this sounds like it should rather be handled in its own effect after this one, which also would make this init method a bit less complex and we nicely separate concerns

File: src/@utils/ocean.ts (hunk 2 above)
Comment (kremalicious) [reply src/@utils/ocean.ts:78]: yup, trivial, with this in `@context/Prices`:

```js
const tokenIds = 'ocean-protocol,ethereum,matic-network'
```

we get this back:

Screen Shot 2022-04-15 at 17 03 52

Requires changing a bit data structure in `usePrices()` as it looks like though

**Comment by kremalicious** [comment]
closing this, as:

- 230 lines of code are too much for this functionality
- nobody uses ETH mainnet in our market, so usefulness is questionable

**Pull request closed without merging**